### PR TITLE
JWT authorization and authentication for UIServer reverse proxy

### PR DESCRIPTION
**Pull request by ignasbernotas**
Closes: #1188

### CODE REVIEW

File: cmd/di.go [hunk 1]
```diff
@@ -26,6 +26,9 @@ import (
 	"github.com/ethereum/go-ethereum/accounts/keystore"
 	"github.com/ethereum/go-ethereum/common"
 	"github.com/ethereum/go-ethereum/ethclient"
+	payment_bindings "github.com/mysteriumnetwork/payment-bindings"
+	"github.com/pkg/errors"
+
```
Comment (soffokl) [inline cmd/di.go]: Extra line

File: core/auth/jwt.go [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package auth
+
+import (
+	"math/rand"
```
Comment (soffokl) [inline core/auth/jwt.go]:
```suggestion
	"crypto/rand"
```

File: core/auth/jwt.go [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+	"time"
+
+	"github.com/dgrijalva/jwt-go"
+	"github.com/pkg/errors"
+)
+
+// JwtAuthenticator contains JWT handling methods
+type JwtAuthenticator struct {
```
Comment (soffokl) [inline core/auth/jwt.go]:
```suggestion
type JWTAuthenticator struct {
```

File: core/auth/jwt.go [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+		return false, err
+	}
+	claims := &jwtClaims{}
+
+	tkn, err := jwt.ParseWithClaims(token, claims, func(token *jwt.Token) (interface{}, error) {
+		return jwtAuth.encryptionKey, nil
+	})
+
+	if tkn == nil || !tkn.Valid {
+		return false, errors.New("Invalid JWT token")
```
Comment (soffokl) [inline core/auth/jwt.go]:
```suggestion
		return false, errors.New("invalid JWT token")
```

File: core/auth/jwt.go [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+}
+
+func (jwtAuth *JwtAuthenticator) setupEncryptionKey() error {
+	key := jwtKey{}
+	err := jwtAuth.storage.GetValue(encryptionKeyBucket, encryptionKeyName, &key)
+	if err != nil {
+		key = generateRandomKey(64)
+		err := jwtAuth.storage.SetValue(encryptionKeyBucket, encryptionKeyName, key)
+		if err != nil {
+			return errors.Wrap(err, "Failed to store JWT encryption key")
```
Comment (soffokl) [inline core/auth/jwt.go]:
```suggestion
			return errors.Wrap(err, "failed to store JWT encryption key")
```

File: ui/server.go [hunk 6]
```diff
@@ -20,19 +20,16 @@ package ui
...
-	"strings"
 	"time"
 
 	log "github.com/cihub/seelog"
 	"github.com/gin-contrib/cors"
 	"github.com/gin-gonic/gin"
 	godvpnweb "github.com/mysteriumnetwork/go-dvpn-web"
-	"github.com/mysteriumnetwork/node/ui/discovery"
 	"github.com/pkg/errors"
+
```
Comment (soffokl) [inline ui/server.go]: Extra line.

File: core/auth/jwt.go [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+)
+
+// JwtAuthenticator contains JWT handling methods
+type JwtAuthenticator struct {
+	storage       Storage
+	encryptionKey []byte
+}
+
+// JWTToken contains token details
+type JWTToken struct {
```
Comment (vkuznecovas) [inline core/auth/jwt.go]: if expanded, this will be jsonwebtokentoken, rename it to JWT.

File: core/auth/jwt.go [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+}
+
+type jwtKey []byte
+
+const expiresIn = 48 * time.Hour
+const encryptionKeyBucket = "jwt"
+const encryptionKeyName = "jwt-encryption-key"
+
+// NewJWTAuthenticator creates a new JWT authentication instance
+func NewJWTAuthenticator(storage Storage) *JwtAuthenticator {
```
Comment (vkuznecovas) [inline core/auth/jwt.go]: I'd redesign this component a bit. You're injecting storage just to store the key - why not actually fetch(or generate and store if it does not exist) the key and pass it as an argument here? No need to inject storage then, and you're not really. You won't need the ensureEncryptionKeyIsSet() method which is a racing anti-pattern anyway. I could now call create/validate token with multiple goroutines and the state would be modified in an upredictable way.

File: core/auth/jwt.go (hunk 2 above)
Comment (ignasbernotas) [reply core/auth/jwt.go]: math/crypto have a different API and in this case pseudo-random works for us.

File: cmd/di.go [hunk 9]
```diff
@@ -22,10 +22,11 @@ import (
 	"net"
 	"path/filepath"
 	"time"
-
```
Comment (vkuznecovas) [inline cmd/di.go]: there should be a space here :)

File: core/auth/jwt.go [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+	claims := &jwtClaims{}
+
+	tkn, err := jwt.ParseWithClaims(token, claims, func(token *jwt.Token) (interface{}, error) {
+		return jwtAuth.encryptionKey, nil
+	})
+
+	if tkn == nil || !tkn.Valid {
+		return false, errors.New("invalid JWT token")
+	}
+	if err != nil {
```
Comment (vkuznecovas) [inline core/auth/jwt.go]: move this up to where the error is being generated. Also - in the current implementation, if the error is not nil, and is not equal to jwt.ErrSignatureInvalid - we return that the token is valid. Is this the intended behaviour?

File: core/auth/jwt.go (hunk 2 above)
Comment (vkuznecovas) [reply core/auth/jwt.go]: does it though? i'd still switch to crypto

**Comment by ignasbernotas** [comment]
@soffokl @vkuznecovas switched to crypto/rand, refactored key generator :)

**Pull request merged**